src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,670 @@
+        }
+
+        private void SessionDismissed(object sender, EventArgs e)
+        {
+            if (sender is IAsyncCompletionSession session)
+            {
+                session.ItemCommitted -= ItemCommitted;
+                session.Dismissed -= SessionDismissed;
+            }
+        }
+
+        public Task<EditorCompletion.FilteredCompletionModel> UpdateCompletionListAsync(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            var mustSetSelection = false;
+
+            var filterText = session.ApplicableToSpan.GetText(data.Snapshot);
+
+            // Check if the user is typing a number. If so, only proceed if it's a number
+            // directly after a <dot>. That's because it is actually reasonable for completion
+            // to be brought up after a <dot> and for the user to want to filter completion
+            // items based on a number that exists in the name of the item. However, when
+            // we are not after a dot (i.e. we're being brought up after <space> is typed)
+            // then we don't want to filter things. Consider the user writing:
+            //
+            //      dim i =<space>
+            //
+            // We'll bring up the completion list here (as VB has completion on <space>).
+            // If the user then types '3', we don't want to match against Int32.
+            if (filterText.Length > 0 && char.IsNumber(filterText[0]))
+            {
+                if (!IsAfterDot(data.Snapshot, session.ApplicableToSpan))
+                {
+                    return Task.FromResult(new EditorCompletion.FilteredCompletionModel(ImmutableArray<EditorCompletion.CompletionItemWithHighlight>.Empty, 0));
+                }
+            }
+
+            // We need to filter if a non-empty strict subset of filters are selected
+            var selectedFilters = data.SelectedFilters.Where(f => f.IsSelected).Select(f => f.Filter).ToImmutableArray();
+            var needToFilter = selectedFilters.Length > 0 && selectedFilters.Length < data.SelectedFilters.Length;
+
+            var initialListOfItemsToBeIncluded = new List<FilterResult>();
+            foreach (var item in data.InitialSortedList)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (needToFilter && ShouldBeFilteredOutOfCompletionList(item, selectedFilters))
+                {
+                    continue;
+                }
+
+                if (MatchesFilterText(item, filterText, data.InitialTrigger.Reason, data.UpdateTrigger.Reason))
+                {
+                    initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: true));
+                }
+                else
+                {
+                    if (data.InitialTrigger.Reason == EditorCompletion.InitialTriggerReason.Deletion ||
```

Comment:
Do this check first before MatchesFilterText since it's a lot cheaper?

Reply:
It seems that MatchFilterText should have another priority over the reasons comparison: if there is a Match, we set matchedFilterText: true even if it is the deletion. So, we cannot swap condition checks.

---
In reply to: [208709031](https://github.com/dotnet/roslyn/pull/29016#discussion_r208709031) [](ancestors = 208709031)